New values and types chapter (#516)
neo4j/docs-cypher: modules/ROOT/pages/values-and-types/spatial.adoc

[[spatial-values]]
= Spatial values

> Cypher has built-in support for handling spatial values (points), which are stored as properties on nodes and relationships in Neo4j databases. 

Reviewer: ```suggestion
Cypher has built-in support for handling spatial values (points), which can be stored as properties on nodes and relationships in Neo4j databases. 
```

> This section begins with an explanation the `Point` type.

Reviewer: ```suggestion
This section begins with an explanation of the `Point` type.
```

> It then proceeds to discuss Cypher's support of Coordinate Reference Systems, and how to work with spatial instants in Cypher, including how spatial points instants work with Cypher indexing. 

Reviewer: ```suggestion
It then proceeds to discuss Cypher's support of Coordinate Reference Systems, and how to work with spatial instants in Cypher, including how spatial point instants work with Cypher indexing. 
```

Finally, it briefly explains comparability and orderability with regard to spatial instants. 

[NOTE]
====
For more information about spatial functions, allowing for the creation and manipulation of spatial values, see the section on xref::functions/spatial.adoc[Spatial functions]. 

For more information about the comparison and ordering of spatial values, see the section on the xref::syntax/operators.adoc#cypher-ordering[ordering and comparison of values].
====

[[spatial-values-point-type]]
== The Point type

Neo4j supports the `Point` type for values of spatial geometry.

Values with the `Point` type have the following characteristics:

* Each point can have either 2 or 3 dimensions. 
This means it contains either 2 or 3 64-bit floating point values, which together are called the _Coordinate_.
* Each point will also be associated with a specific xref::values-and-types/spatial.adoc#cypher-spatial-crs[Coordinate Reference System] (CRS) that determines the meaning of the values in the _Coordinate_.
* Instances of `Point` and lists of `Point` can be assigned to node and relationship properties.

> * Nodes with `Point` or `List(Point)` properties can be indexed using a xref::indexes-for-search-performance.adoc#administration-indexes-create-a-node-point-index

Reviewer: ```suggestion
* Nodes and Relationships with `Point` or `List(Point)` properties can be indexed using a xref::indexes-for-search-performance.adoc#administration-indexes-create-a-node-point-index
```

Reviewer: Also I think the link shouldn't have the administration part in it?

[point index]. 

> This is true for all CRS (and for both 2D and 3D).

Reviewer: I have a hard time reading CRS as plural 😅

* The xref::functions/spatial.adoc#functions-distance[distance function] will work on points in all CRS and in both 2D and 3D, but only if the two points have the same CRS (and therefore also same dimension).

[[spatial-values-crs]]
== Coordinate Reference Systems

Four Coordinate Reference Systems (CRS) are supported, each of which falls within one of two types: _geographic coordinates_, modeling points on the earth, or _Cartesian coordinates_, modeling points in euclidean space:

Data within different coordinate systems are entirely incomparable, and cannot be implicitly converted from one to the other.

> This is true even if they are both Cartesian or both geographic. 

Reviewer: ```suggestion
This is true even if they are both Cartesian or both geographic but of a different dimension. 
```

 ** Specifying this CRS can be done using either the name 'wgs-84-3d' or the SRID 4979 as described in xref::functions/spatial.adoc#functions-point-wgs84-3d[Point(WGS-84-3D)].

[[spatial-values-converting-coordinates]]
=== Converting coordinate units

The units of the `latitude` and `longitude` fields are in decimal degrees, and need to be specified as floating point numbers using Cypher literals.
It is not possible to use any other format, such as 'degrees, minutes, seconds'.
The units of the `height` field are in meters.
When geographic points are passed to the xref::functions/spatial.adoc#functions-distance[distance function], the result will always be in meters.
If the coordinates are in any other format or unit than those supported, it is necessary to explicitly convert them. 

For example, if the incoming `$height` is a string field in kilometers, it would be necessary to add `height: toFloat($height) * 1000` to the query.
Likewise if the results of the `distance` function are expected to be returned in kilometers, an explicit conversion is required.
The below query is an example of this conversion: 

.Query

> [source, cypher, indent=0]

Reviewer: Do we have style guide rules around map key:value spacing? I personally feel like there should be a space after the colon, so just wanted to check

Author: We do, and there should :) 
https://neo4j.com/docs/cypher-manual/current/styleguide/#cypher-styleguide-spacing

 ** A 3D point in the _Cartesian_ CRS is specified with a map containing `x`, `y` and `z` coordinate values
 ** Specifying this CRS can be done using either the name 'cartesian-3d' or the SRID 9157 as described in xref::functions/spatial.adoc#functions-point-cartesian-3d[Point(Cartesian-3D)]

The units of the `x`, `y`, and `z` fields are unspecified. 
This  means that when two Cartesian points are passed to the `distance` function, the resulting value will be in the same units as the original coordinates.

> This is true for both 2D and 3D points, as the _pythagoras_ equation used is generalized to any number of dimensions. 

Reviewer: Should Pythagoras be capitalised as it is technically a name?

However, just as you cannot compare geographic points to Cartesian points, you cannot calculate the distance between a 2D point and a 3D point.
If you need to do that, explicitly transform the one type into the other.
For example:

> .Query

Reviewer: bad and good feel funny :P not sure if it matters

| `instant.srid` | The internal Neo4j ID for the CRS | Integer | One of `4326`, `4979`, `7203`, `9157` | {check-mark} | {check-mark} | {check-mark} | {check-mark}

> |===

Reviewer: It bothers me that latitude comes before longitude when the description says first and second in it 😅

[[spatial-values-point-index]]
== Spatial values and indexes
// POINT INDEX new in Neo4j 5.0

> If there is a RANGE or POINT index on a particular `:Label(property) combination`, and a spatial point is assigned to that property on a node with that label, the node will be indexed.

Reviewer: Rels can also have indexes for point :)
